library/math/num.lisp

```diff
@@ -157,29 +157,47 @@
 (%define-overflow-handler %handle-64bit-overflow 64)
 (%define-overflow-handler %handle-fixnum-overflow #.+fixnum-bits+)
 
+;;; 
```

Comment:
Shouldn't there also be overflow handlers for unsigned values?

Reply:
Unsigned values don’t overflow

Reply:
If the result is supposed to be an unsigned 8-bit Integer, and it evaluates to > 127, wouldn't that be an overflow? 

Even if it wraps around and modulos it or somewhat, that still seems like potentially problematic behavior.

Reply:
A quote from the blog post I shared above;

“ An unsigned 8-bit integer can store any value between 0 and 255…”

Reply:
Unsigned Ints wrap around but don’t overflow. https://stackoverflow.com/questions/9193880/overflowing-of-unsigned-int

Reply:
Whoops I got turned around, I was thinking in signed for a second, but still... 

The blog post you shared talks about the different ways to handle overflow and underflow, and talks about how wrapping around is less expensive but can cause problems, same with saturation. 

For example, a U8 derived from `(+ 255 1)` should trigger an overflow error, since it can't return 256 without changing to a larger Integer type.
The other two options return questionable results:
With wrap around: `(+ 255 1)` => `1`.
With saturation `(+ 255 1)` => `255`.

The point of the handler is to return an error so you don't get the **wrong** answer somewhere in your code.

Reply:
I see what you’re saying.

Here’s the  original issue and it says to signal an error if unsigned numbers underflow. So I a was assuming I didn’t need to check whether they overflowed. https://github.com/coalton-lang/coalton/issues/921

Reply:
That makes sense, I was looking at this issue: #963 which is about signed numbers anyways.

I guess the jury's still out.

Reply:
You are right about wrap around in Coalton-

In SBCL, this errors because 256 is too big for unsigned 8-bit:
```
(declaim (ftype (function ((unsigned-byte 8) (unsigned-byte 8)) (unsigned-byte 8)) add-8bit))

(defun add-8bit (x y)
  (+ x y))

(add-8bit 255 1)
```
But in Coalton, this evaluates to 0:
```
(coalton-toplevel

  (declare add-8bit (U8 -> U8 -> U8))
  (define (add-8bit x y)
    (+ x y)))

(coalton (add-8bit 255 1))
```

Reply:
So is my code correct?

Reply:
So for the four cases, unsigned and signed, overflow and underflow, this is what I'm currently getting with your code, let me know if it's consistent with your results:

For signed 8-bit, it's throwing an overflow error for both overflow and underflow. For unsigned, it's wrapping around in both directions, which seems to be what you were trying to change. 

I did check, and these happen to be the same results for the main branch as well.

```
(coalton-toplevel
  (declare add-8bit (U8 -> U8 -> U8))
  (define (add-8bit x y)
    (+ x y))

  (declare sub-8bit (U8 -> U8 -> U8))
  (define (sub-8bit x y)
    (- x y))

  (declare add-I8 (I8 -> I8 -> I8))
  (define (add-I8 x y)
    (+ x y))
  
  (declare sub-I8 (I8 -> I8 -> I8))
  (define (sub-I8 x y)
    (- x y)))
```
```
COALTON-USER> (sub-I8 -50 100)

 Signed value overflowed 8 bits.
```
```
COALTON-USER> (add-I8 50 100)

 Signed value overflowed 8 bits.
```
```
COALTON-USER> (add-8bit 255 1)
0
```
```
COALTON-USER> (sub-8bit 1 10)
247
```

Reply:
Thanks - I’ll look into it